File: app/src/test/java/org/thoughtcrime/securesms/jobmanager/migrations/SendReadReceiptsJobMigrationTest.java
package org.thoughtcrime.securesms.jobmanager.migrations;

import org.junit.Test;
import org.thoughtcrime.securesms.database.MessageTable;
import org.thoughtcrime.securesms.jobmanager.JsonJobData;
import org.thoughtcrime.securesms.jobmanager.JobMigration;
import org.thoughtcrime.securesms.jobs.SendReadReceiptJob;
import org.thoughtcrime.securesms.recipients.RecipientId;

import java.util.ArrayList;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class SendReadReceiptsJobMigrationTest {

  private final MessageTable                 mockDatabase = mock(MessageTable.class);
  private final SendReadReceiptsJobMigration testSubject  = new SendReadReceiptsJobMigration(mockDatabase);

  @Test
  public void givenSendReadReceiptJobDataWithoutThreadIdAndThreadIdFound_whenIMigrate_thenIInsertThreadId() {
    // GIVEN
    SendReadReceiptJob   job     = new SendReadReceiptJob(1, RecipientId.from(2), new ArrayList<>(), new ArrayList<>());
    JobMigration.JobData jobData = new JobMigration.JobData(job.getFactoryKey(),
                                                            "asdf",
                                                            new JsonJobData.Builder()
                                                                    .putString("recipient", RecipientId.from(2).serialize())
                                                                    .putLongArray("message_ids", new long[]{1, 2, 3, 4, 5})
                                                                    .putLong("timestamp", 292837649).serialize());
    when(mockDatabase.getThreadIdForMessage(anyLong())).thenReturn(1234L);

    // WHEN
    JobMigration.JobData result = testSubject.migrate(jobData);
    JsonJobData          data   = JsonJobData.deserialize(result.getData());

    // THEN
    assertEquals(1234L, data.getLong("thread"));
    assertEquals(RecipientId.from(2).serialize(), data.getString("recipient"));
    assertTrue(data.hasLongArray("message_ids"));
    assertTrue(data.hasLong("timestamp"));
  }

  @Test
  public void givenSendReadReceiptJobDataWithoutThreadIdAndThreadIdNotFound_whenIMigrate_thenIGetAFailingJob() {
    // GIVEN
    SendReadReceiptJob   job     = new SendReadReceiptJob(1, RecipientId.from(2), new ArrayList<>(), new ArrayList<>());
    JobMigration.JobData jobData = new JobMigration.JobData(job.getFactoryKey(),
                                                            "asdf",
                                                            new JsonJobData.Builder()
                                                                .putString("recipient", RecipientId.from(2).serialize())
                                                                .putLongArray("message_ids", new long[]{})
                                                                .putLong("timestamp", 292837649).serialize());
    when(mockDatabase.getThreadIdForMessage(anyLong())).thenReturn(-1L);

    // WHEN
    JobMigration.JobData result = testSubject.migrate(jobData);

    // THEN
    assertEquals("FailingJob", result.getFactoryKey());
  }

  @Test
  public void givenSendReadReceiptJobDataWithThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    SendReadReceiptJob   job     = new SendReadReceiptJob(1, RecipientId.from(2), new ArrayList<>(), new ArrayList<>());
    JobMigration.JobData jobData = new JobMigration.JobData(job.getFactoryKey(), "asdf", job.serialize());

    // WHEN
    JobMigration.JobData result = testSubject.migrate(jobData);

    // THEN
    assertEquals(jobData, result);
  }

  @Test
  public void givenSomeOtherJobDataWithThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    JobMigration.JobData jobData = new JobMigration.JobData("SomeOtherJob", "asdf", new JsonJobData.Builder().putLong("thread", 1).serialize());

    // WHEN
    JobMigration.JobData result = testSubject.migrate(jobData);

    // THEN
    assertEquals(jobData, result);
  }

  @Test
  public void givenSomeOtherJobDataWithoutThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    JobMigration.JobData jobData = new JobMigration.JobData("SomeOtherJob", "asdf", new JsonJobData.Builder().serialize());

    // WHEN
    JobMigration.JobData result = testSubject.migrate(jobData);

    // THEN
    assertEquals(jobData, result);
  }


}

File: app/src/test/java/org/thoughtcrime/securesms/jobmanager/migrations/SendReadReceiptsJobMigrationTest.kt
package org.thoughtcrime.securesms.jobmanager.migrations

import io.mockk.every
import io.mockk.mockk
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.thoughtcrime.securesms.database.MessageTable
import org.thoughtcrime.securesms.jobmanager.JobMigration.JobData
import org.thoughtcrime.securesms.jobmanager.JsonJobData
import org.thoughtcrime.securesms.jobs.SendReadReceiptJob
import org.thoughtcrime.securesms.recipients.RecipientId

class SendReadReceiptsJobMigrationTest {
  private val mockDatabase = mockk<MessageTable>()
  private val testSubject = SendReadReceiptsJobMigration(mockDatabase)

  @Test
  fun givenSendReadReceiptJobDataWithoutThreadIdAndThreadIdFound_whenIMigrate_thenIInsertThreadId() {
    // GIVEN
    val job = SendReadReceiptJob(1, RecipientId.from(2), ArrayList(), ArrayList())
    val jobData = JobData(
      factoryKey = job.factoryKey,
      queueKey = "asdf",
      maxAttempts = -1,
      lifespan = -1,
      data = JsonJobData.Builder()
        .putString("recipient", RecipientId.from(2).serialize())
        .putLongArray("message_ids", longArrayOf(1, 2, 3, 4, 5))
        .putLong("timestamp", 292837649).serialize()
    )
    every { mockDatabase.getThreadIdForMessage(any()) } returns 1234L

    // WHEN
    val result = testSubject.migrate(jobData)
    val data = JsonJobData.deserialize(result.data)

    // THEN
    assertEquals(1234L, data.getLong("thread"))
    assertEquals(RecipientId.from(2).serialize(), data.getString("recipient"))
    assertTrue(data.hasLongArray("message_ids"))
    assertTrue(data.hasLong("timestamp"))
  }

  @Test
  fun givenSendReadReceiptJobDataWithoutThreadIdAndThreadIdNotFound_whenIMigrate_thenIGetAFailingJob() {
    // GIVEN
    val job = SendReadReceiptJob(1, RecipientId.from(2), ArrayList(), ArrayList())
    val jobData = JobData(
      factoryKey = job.factoryKey,
      queueKey = "asdf",
      maxAttempts = -1,
      lifespan = -1,
      data = JsonJobData.Builder()
        .putString("recipient", RecipientId.from(2).serialize())
        .putLongArray("message_ids", longArrayOf())
        .putLong("timestamp", 292837649).serialize()
    )
    every { mockDatabase.getThreadIdForMessage(any()) } returns -1L

    // WHEN
    val result = testSubject.migrate(jobData)

    // THEN
    assertEquals("FailingJob", result.factoryKey)
  }

  @Test
  fun givenSendReadReceiptJobDataWithThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    val job = SendReadReceiptJob(1, RecipientId.from(2), ArrayList(), ArrayList())
    val jobData = JobData(job.factoryKey, "asdf", -1, -1, job.serialize())

    // WHEN
    val result = testSubject.migrate(jobData)

    // THEN
    assertEquals(jobData, result)
  }

  @Test
  fun givenSomeOtherJobDataWithThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    val jobData = JobData("SomeOtherJob", "asdf", -1, -1, JsonJobData.Builder().putLong("thread", 1).serialize())

    // WHEN
    val result = testSubject.migrate(jobData)

    // THEN
    assertEquals(jobData, result)
  }

  @Test
  fun givenSomeOtherJobDataWithoutThreadId_whenIMigrate_thenIDoNotReplace() {
    // GIVEN
    val jobData = JobData("SomeOtherJob", "asdf", -1, -1, JsonJobData.Builder().serialize())

    // WHEN
    val result = testSubject.migrate(jobData)

    // THEN
    assertEquals(jobData, result)
  }
}
